diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,410 @@
+
+    _t1 = _get_t(i, ts)
+    next_made_jump = _t1 <= t1
+    _t1 = jnp.where(next_made_jump, _t1, t1)
+    return _t1, next_made_jump
+
+
+def _find_index(t: RealScalarLike, ts: Optional[Array]) -> IntScalarLike:
+    if ts is None:
+        return 0
+
+    ts = upcast_or_raise(
+        ts,
+        t,
+        "`PIDController.step_ts`",
+        "time (the result type of `t0`, `t1`, `dt0`, `SaveAt(ts=...)` etc.)",
+    )
+    return jnp.searchsorted(ts, t, side="right")
+
+
+def _revisit_rejected(
+    t0: RealScalarLike,
+    t1: RealScalarLike,
+    i_rjct: IntScalarLike,
+    rjct_buff: Optional[Array],
+) -> RealScalarLike:
+    if rjct_buff is None:
+        return t1
+    _t1 = _get_t(i_rjct, rjct_buff)
+    _t1 = jnp.minimum(_t1, t1)
+    return _t1
+
+
+# EXPLANATION OF STEP_TS AND JUMP_TS
+# -----------------------------------
+# The `step_ts` and `jump_ts` are used to force the solver to step to certain times.
+# They mostly act in the same way, except that when we hit an element of `jump_ts`,
+# the controller must return `made_jump = True`, so that the diffeqsolve function
+# knows that the vector field has a discontinuity at that point. In addition, the
+# exact time of the jump will be skipped using jnp.prevbefore and jnp.nextafter.
+# So now to the explanation of the two (we will use `step_ts` as an example, but the
+# same applies to `jump_ts`):
+#
+# If `step_ts` is not None, we assume it is a sorted array of times.
+# At the start of the run, the init function finds the smallest index `i_step` such
+# that `step_ts[i_step] > t0`. At init and after each step of the solver, the
+# controller will propose a step t1_next, and we will clip it to
+# `t1_next = min(t1_next, step_ts[i_step])`.
+# At the start of the next step, if the step ended at t1 == step_ts[i_step] and
+# if the controller decides to keep the step, then this time has been successfully
+# stepped to and we increment `i_step` by 1.
+# We use a convenience function _get_t(i, ts) which returns ts[i] if i < len(ts) and
+# infinity otherwise.
+
+# EXPLANATION OF REVISITING REJECTED STEPS
+# ----------------------------------------
+# We use a "stack" of rejected steps, composed of a buffer `rjct_buff` of length
+# `rejected_step_buffer_len` and a counter `i_rjct`. The "stack" are all the items
+# in `rjct_buff[i_rjct:]` with `rjct_buff[i_rjct]` being the top of the stack.
+# When `i_rjct == rejected_step_buffer_len`, the stack is empty.
+# At the start of the run, `i_rjct = rejected_step_buffer_len`. Each time a step is
+# rejected `i_rjct -=1` and `rjct_buff[i_rjct] = t1`. Each time a step ends at
+# `t1 == rjct_buff[i_rjct]`, we increment `i_rjct` by 1 (even if the step was
+# rejected, in which case we will re-add `t1` to the stack immediately).
+# We clip the next step to `t1_next = min(t1_next, rjct_buff[i_rjct])`.
+# If `i_rjct < 0` then an error is raised.
+
+
+class JumpStepWrapper(
+    AbstractStepSizeController[_JumpStepState[_ControllerState], _Dt0]
+):
```

Comment:
I think we should make this -- and the wrapped `.controller` -- specifically be `AbstractAdaptiveStepSizeController`. We do have a couple of places that check for adaptivity via `isinstance(...., AbstractAdaptiveStepSizeController)`.

(More generally this is something where in retrospect I probably shouldn't have lifted certain properties, like adaptivity, into the type system, as it makes wrapping harder. At least in this case I think it's fine, I don't really anticipate use cases for this with fixed step size controllers -- do you?)

Reply:
I agree, probably making it adaptive-only makes sense.